### PR TITLE
perf(promql): avoid per-window allocations in simple range functions

### PR DESCRIPTION
**Pull request by discord9**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

No prerequisite PR or issue.

## What's changed and what's your intention?

Reduce per-window allocations in `count_over_time`, `last_over_time`, `present_over_time`, and `absent_over_time` without changing their results or arithmetic complexity.

These functions already do O(1) work per window, but the generic range-function wrapper creates timestamp and value Arrow slice objects and downcasts them for every window. This PR adds an optional `evaluator` path to `#[range_fn]` and uses it for these four functions. The specialized evaluator validates the inputs, downcasts the backing arrays once, and reads each window's existing offset/length directly. Other range functions retain the existing macro expansion.

Preserved behavior includes empty windows, independent timestamp/value offsets, sliced backing arrays and dictionary keys, range-validation order and error messages, zero-window inputs, and UDF metadata. The validation-only timestamp type check uses `Any::is::<TimestampMillisecondArray>()`; value arrays are still downcast for access.

The review follow-up also corrects pre-existing NULL handling in these four functions: NULL fields represent missing samples, so count includes only valid samples, last skips trailing NULLs, and present/absent depend on whether any valid sample exists. All-NULL windows produce no count/last/present sample and produce 1 for absent. Valid NaNs are not treated as NULL. The existing no-NULL evaluator is selected once per batch and remains unchanged; nullable inputs scan validity without per-window Arrow allocations. This intentionally fixes NULL query results; there are no API signatures, schema, or persisted-format changes.

Fixed-expectation UDF tests and the generated `range_presence_null` SQL regression cover mixed, trailing-NULL, all-NULL, and empty windows. The SQL case passed in both standalone and distributed modes. Other range functions' NULL handling is outside this PR.

### Related smoothing allocation cleanup

Also remove the redundant `values.to_vec()` in `double_exponential_smoothing`. The recurrence only reads its input, so borrowing the original slice avoids a per-window allocation and copy without changing guards, initialization, or floating-point operation order. This is allocation/copy removal, not an incremental smoothing algorithm.

The addition includes an old-copy differential oracle, four prepared-input benchmarks (short/long and overlapping/non-overlapping windows), and `tests/perf/query_cases/promql_smoothing_copy/case.toml` with seven queries and unchanged 10% regression thresholds.

The standalone [nightly run 34479645824](https://github.com/GreptimeTeam/greptimedb/actions/runs/34479645824) passed all seven gates, with 63 successful samples per target, zero validation errors, successful teardown, and all retained numeric-control responses matching `240`, `480`, and `720`. Smoothing query medians decreased by 1.4%–4.9%, but the selector also decreased by 3.4%; this does **not** establish an isolated end-to-end speedup from this cleanup. It only found no threshold regression in these scenarios.

The results below verify the final revision including the NULL follow-up. The performance fixtures contain no NULL samples and validate the unchanged no-NULL fast path; NULL correctness is covered separately by fixed-expectation unit tests and standalone/distributed SQL regression tests.

### Tests and performance regression coverage

- Add differential tests against the existing slice kernels, including null raw payloads, NaNs, infinities, signed zero, empty/singleton/overlapping/repeated/backward windows, sliced inputs, and error paths.
- Add eight prepared-input microbenchmarks: four functions with 4,077 overlapping windows and with eight sparse windows. Input construction is outside the timed loop; existing benchmarks remain unchanged.
- Add `tests/perf/query_cases/promql_simple_range_allocation/case.toml`: 128 series × 780 samples at 15-second cadence, short/long windows for all four functions, a selector control, and a Prometheus HTTP numeric control. All ten queries retain a 10% latency-regression threshold. The absent cases exercise nonempty input producing an empty result.

### Measured allocation reduction

A diagnostic executable linked against the optimized baseline/candidate libraries measured allocations during prepared UDF calls, excluding input construction. The diagnostic allocator is not part of this PR.

| Layout | Functions | Allocation calls, base → candidate | Reallocation calls, base → candidate |
| --- | --- | ---: | ---: |
| 4,077 overlapping windows | count / last / present | 8,173 → 19 | 30 → 20 |
| 4,077 overlapping windows | absent | 8,175 → 21 | 30 → 20 |
| Eight sparse windows | count / last / present | 35 → 19 | 3 → 2 |
| Eight sparse windows | absent | 37 → 21 | 3 → 2 |

This removes exactly two allocations per window. Batch-level input validation and output allocations remain. Single-call counts and averages over 100 calls agree. For count/last/present in the overlapping layout, allocation-request bytes decrease from 914,376 to 66,296, and reallocation-request bytes from 261,888 to 130,944; these are requested sizes, not live or peak memory.

### Nightly end-to-end results

[Final query regression run 34563492520](https://github.com/GreptimeTeam/greptimedb/actions/runs/34563492520), `nightly` profile, compares base `0f625a7e922074692767aa9382619deda7aa2b32` with candidate `0a69baf5707f6244794737ce7ab47f01c2639598` including the NULL follow-up. Each query uses three warmups and nine measured iterations.

| Query | Base median ms | Candidate median ms | Change |
| --- | ---: | ---: | ---: |
| `selector_2h15s_control` | 25.11 | 23.29 | -7.3% |
| `count_over_time_1m` | 16.19 | 14.03 | -13.3% |
| `count_over_time_1h` | 16.26 | 14.19 | -12.7% |
| `last_over_time_1m` | 16.22 | 14.52 | -10.5% |
| `last_over_time_1h` | 16.43 | 14.68 | -10.6% |
| `present_over_time_1m` | 16.97 | 14.86 | -12.4% |
| `present_over_time_1h` | 16.88 | 14.58 | -13.6% |
| `absent_over_time_1m_empty` | 15.07 | 12.22 | -18.9% |
| `absent_over_time_1h_empty` | 15.01 | 12.01 | -20.0% |
| `prom_http_last_over_time_1h_numeric_control` | 9.22 | 9.16 | -0.7% |
| `selector_2h_15s_control` | 26.15 | 24.36 | -6.8% |
| `double_exponential_smoothing_1m_15s` | 18.88 | 19.33 | +2.4% |
| `double_exponential_smoothing_5m_15s` | 20.40 | 20.82 | +2.1% |
| `double_exponential_smoothing_1h_15s` | 42.95 | 43.46 | +1.2% |
| `double_exponential_smoothing_1h_5m_overlap` | 15.03 | 14.43 | -4.0% |
| `double_exponential_smoothing_1h_1h_nonoverlap` | 13.07 | 12.99 | -0.6% |
| `double_exponential_smoothing_1h_host0000_numeric_control` | 9.52 | 9.51 | -0.2% |

All 17 gates pass, with 153 successful measured samples per target (306 total), zero validation errors, and successful teardown. Both targets' nine retained HTTP responses in each case return the expected `240`, `480`, and `720` at the three hourly timestamps. The performance runner does not automatically assert numerical equality; these responses were checked separately.

The four simple functions show 10.5%–20.0% lower medians in this run, while their selector control changes by −7.3%; not all observed improvement can be attributed to the optimization. Smoothing query changes range from −4.0% to +2.4% (excluding controls), while its selector changes by −6.8%. No threshold regression was observed, but this is not evidence of an isolated smoothing speedup. Both cases retain their 10% regression gates. NULL correctness was verified separately in the unit and SQL regression tests; these non-NULL performance fixtures do not measure nullable-window performance.

### Verification

- `cargo check -p promql -p common-macro` — passed.
- `cargo nextest run -p promql -p common-macro` — 226 passed, 0 skipped (including the NULL fix).
- `cargo clippy -p promql -p common-macro --all-targets -- -D warnings` — passed.
- `cargo fmt --all -- --check` — passed.
- `cargo run --bin sqlness-runner -- bare --bins-dir <candidate-binary-directory> -t range_presence_null` — passed in standalone and distributed modes; generated results checked against the expected NULL semantics.
- `cargo bench -p promql --bench bench_main -- presence_range_fn --warm-up-time 1 --measurement-time 2 --sample-size 30 --noplot` — completed for both revisions using identical benchmark code. Local timing is noisy; the table above uses nightly end-to-end results instead.
- `cargo build -p cmd --bin query_perf_fixture --bin query_regression_runner --features dev-tools` — passed.
- `query_perf_fixture plan --case tests/perf/query_cases/promql_simple_range_allocation/case.toml` — passed using the built candidate tool.
- `taplo fmt --check --no-auto-config tests/perf/query_cases/promql_simple_range_allocation/case.toml` — passed, with the file actually checked.
- The six Python suites in the query-regression workflow's `test-tooling` job — 20 tests passed.
- Commit/push checks: `hawkeye check`, formatting, typos, and SNAFU checks passed.

Validation was targeted to the changed crates and performance tooling; the full workspace test/clippy suites were not run. No existing tests were removed or thresholds relaxed.

## PR Checklist
Please convert it to a draft if some of the following conditions are not met.

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.
- [x] API changes are backward compatible.
- [x] Schema or data changes are backward compatible.
- [ ] This PR needs to be backported to release branches, and I have added the `backport-<target>` labels (e.g. `backport-v1.3` targets `release/v1.3`).

### CODE REVIEW

File: src/promql/src/functions/aggr_over_time.rs [hunk 1]
```diff
@@ -26,6 +26,103 @@ use datatypes::arrow::datatypes::DataType;
...
+
+    if timestamp_ranges.is_empty() {
+        return Ok(ColumnarValue::Array(Arc::new(Float64Array::from_iter(
+            std::iter::empty::<Option<f64>>(),
+        ))));
+    }
+
+    // The generic range-function wrapper downcasts both arrays for each window.
+    // Do it once after retaining its zero-window behavior above.
+    timestamp_ranges
```
Comment (killme2008) [inline src/promql/src/functions/aggr_over_time.rs]: Why we call `downcast_ref` here? Just for side effects?

**Review by killme2008 — APPROVED** [review]
LGTM

File: src/promql/src/functions/aggr_over_time.rs (hunk 1 above)
Comment (killme2008) [reply src/promql/src/functions/aggr_over_time.rs]: I see your point. I think

```rust
let _ = ...
```

is a clearer way to express this behavior.

**Comment by killme2008** [comment]
One item in the "Preserved behavior" list is worth flagging: *physical-slot counting* and *raw last-slot values even under a null bitmap*. That behavior is a bug. Preserving it is fine for this PR's scope, but it shouldn't be read as a contract.

In GreptimeDB a NULL field value means the series has no sample at that timestamp. In PromQL terms the sample doesn't exist and has to be skipped. Every float range function reads the Float64 buffer and ignores the null bitmap, so the padding (0.0 in practice) is treated as a real sample.

Repro on `38577c8`:

```sql
CREATE TABLE t2 (ts TIMESTAMP(3) TIME INDEX, host STRING PRIMARY KEY, val DOUBLE);
INSERT INTO t2 VALUES (0,'a',1.0),(1000,'a',NULL),(2000,'a',NULL),(3000,'a',4.0);
TQL EVAL (3, 3, '1s') count_over_time(t2[4s]);
```

| function | actual | PromQL |
| --- | --- | --- |
| `count_over_time` | 4 | 2 |
| `avg_over_time` | 1.25 | 2.5 |
| `last_over_time`, trailing NULL | 0.0 | last non-null |
| `present_over_time`, all-NULL window | 1 | no result |
| `rate` | 1.25 | 1.0 |
| `changes` | 2 | 1 |
| `quantile_over_time(0.5, ...)` | 0.5 | 2.5 |

`deriv` and `predict_linear` are already correct — `linear_regression_slices` guards on `has_nulls = values.null_count() > 0` and skips null slots. The pattern already exists in the crate.

Since you're already touching these four, fixing them here is cheap:

- `Count` → valid count, 0 → `None`
- `Last` → walk back to the first non-null
- `Present` / `Absent` → based on the valid count

Check `values.null_count() == 0` once outside the loop and pick the fn pointer from that. Tables without NULLs keep exactly the path you benchmarked, so `promql_simple_range_allocation` shouldn't move.

Your differential test holds either way: it asserts evaluator == slice kernel rather than fixed values, and `special_ranges()` already has a real null slot at index 5. Both sides just need to change together.

Happy to take it in a separate PR instead if you'd rather keep this one behavior-neutral — I'm fixing the rest (`rate` / `changes` / `resets` / `idelta` / `quantile_over_time`) anyway.

Unrelated to your change but in the same file: `stdvar_over_time` (`aggr_over_time.rs:145` on main) and `stddev_over_time` (`:178`) call `value.unwrap()` on the `Option<f64>` iterator, so any NULL in the window panics. tokio swallows the panic and the query returns 200 with that series silently missing. Pre-existing — I'll handle it.

File: src/promql/src/functions/aggr_over_time.rs (hunk 1 above)
Comment (discord9) [reply src/promql/src/functions/aggr_over_time.rs]: Addressed in 0a69baf5707: this is now an explicit `assert!(timestamp_ranges.values().as_any().is::<TimestampMillisecondArray>())`, since only type validation is needed. It remains after the zero-window return; the value array still uses `downcast_ref()` because its data is read. Checks, clippy, and all 226 tests pass.

**Comment by discord9** [comment]
@killme2008 Thanks for flagging the NULL semantics. Fixed these four functions in 0a69baf5707: count skips NULLs, last finds the last valid sample, and present/absent test for any valid sample. The no-NULL closures are unchanged and selected once per batch; valid NaNs remain samples. I also corrected the PR description so the old NULL behavior is not presented as a contract.

Added fixed-expectation UDF tests (not just a differential oracle) and `range_presence_null.sql` covering mixed, trailing-NULL, all-NULL, and empty windows. The SQL regression passes in both standalone and distributed modes; 226 tests and clippy pass.

The final nightly run https://github.com/GreptimeTeam/greptimedb/actions/runs/34563492520 passes all 17 gates, with 306 successful samples, zero validation errors, and successful teardown. Its fixtures validate the no-NULL performance path; NULL correctness is covered by the separate unit/SQL tests. Other range functions are left untouched to avoid overlapping your fixes.
